### PR TITLE
Allow for an in-memory database to be used during testing

### PR DESCRIPTION
**Pull request by thomaseizinger**

### CODE REVIEW

File: cnd/src/db/mod.rs [hunk 1]
```diff
@@ -28,66 +32,116 @@ impl Sqlite {
...
-            .context("failed to determine default path for database ")?;
-        let db = Sqlite { db: file };
+    pub fn new(location: Location<'_>) -> anyhow::Result<Self> {
+        let db = match location {
+            Location::OnDisk(path) => {
+                if path == Path::new(":memory:") {
+                    anyhow::bail!("use Location::InMemory if you want an in-memory database!")
+                }
+
+                ensure_folder_tree_exists(path)?;
```
Comment (tcharding) [inline cnd/src/db/mod.rs]:
```suggestion
                ensure_directory_tree_exists(path)?;
```
It's bad enough only half the team are Linux hackers without using OSX terminology too :)

File: cnd/src/db/mod.rs [hunk 2]
```diff
@@ -28,66 +32,116 @@ impl Sqlite {
...
-    }
-
     fn connect(&self) -> anyhow::Result<SqliteConnection> {
-        let db = &self.db;
-        let database_url = db
-            .to_str()
-            .with_context(|| format!("{} is not a valid path", db.display()))?;
-        let connection = SqliteConnection::establish(&database_url)?;
-        Ok(connection)
+        Ok(SqliteConnection::establish(&self.uri)?)
```
Comment (tcharding) [inline cnd/src/db/mod.rs]: This is nice, thanks for the clean up.

File: cnd/src/db/mod.rs [hunk 3]
```diff
@@ -8,18 +8,22 @@ mod serialization_format_stability_tests;
...
 
 /// This module provides persistent storage by way of Sqlite.
 
 #[derive(Debug, Clone)]
 pub struct Sqlite {
-    db: PathBuf,
+    uri: String,
+}
+
+/// Defines the storage location of our SQLite database
```
Comment (tcharding) [inline cnd/src/db/mod.rs]:
```suggestion
/// Defines the storage location of our SQLite database.
```

File: cnd/src/db/mod.rs [hunk 4]
```diff
@@ -28,66 +32,116 @@ impl Sqlite {
...
-        let temp_file_path = temp_file.into_temp_path().to_path_buf();
 
-        let db = Sqlite::new(Some(temp_file_path));
+        temp_file.into_temp_path().to_path_buf()
+    }
+
+    #[test]
+    fn can_create_a_new_temp_db() {
+        let path = temp_db();
+
```
Comment (tcharding) [inline cnd/src/db/mod.rs]:
```suggestion
```

**Comment by thomaseizinger** [comment]
> Sweet tests. I pushed a refactor patch to use PathBuf instead of the tempfile returned type. Hope you don't mind this process. Feel free to remove the patch if you don't want it.

Thanks @tcharding :)

File: cnd/src/db/mod.rs (hunk 1 above)
Comment (D4nte) [reply cnd/src/db/mod.rs]: Today I learned something.

**Pull request merged**